### ISSUE TITLE
MissingOuputException running Daijin

### ISSUE THREAD

**Issue by NJeanray** [EI-CoreBioinformatics/mikado#325]
Dear Team,

I'm just starting experiencing Mikado (version 2.0rc2) and I'm encountering an issue running Daijin : 

> Waiting at most 60 seconds for missing files.
MissingOutputException in line 997 of /softs/anaconda3/lib/python3.7/site-packages/Mikado/daijin/assemble.smk:
Job completed successfully, but some output files are missing. Missing files after 60 seconds:
Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.tab
This might be due to filesystem latency. If that is the case, consider to increase the wait time with --latency-wait.
  File "/softs/anaconda3/lib/python3.7/site-packages/snakemake/executors/__init__.py", line 544, in handle_job_success
  File "/softs/anaconda3/lib/python3.7/site-packages/snakemake/executors/__init__.py", line 231, in handle_job_success

This is very strange as the file mentionned as "missing" exists : 

> lrwxrwxrwx 1 NJEANRAY Utilisa. du domaine 77 Aug 28 16:25 portcullis_hisat-Patient_24-0.pass.junctions.tab -> ../portcullis_hisat-Patient_24-0/3-filt/hisat-Patient_24-0.pass.junctions.tab

My sample_sheet is really simple : 

> Reads/SNIPr_01_03_24_43_2009_9_16_VI_RP_1_S5_ME_R1_001.fastq.gz	Reads/SNIPr_01_03_24_43_2009_9_16_VI_RP_1_S5_ME_R2_001.fastq.gz	Patient_24	fr-firststrand	False

And here are the command lines I launched : 

` daijin configure --scheduler SLURM   -al hisat -as class stringtie   --scoring HISTORIC/hsapiens_scoring.yaml -m permissive --sample-sheet sample_sheet.tsv  --threads 4 --genome Reference/Homo_sapiens.GRCh38.dna.toplevel.fa  --queue NJE_24 -o daijin.yaml`

`daijin assemble --cores 40 daijin.yaml `

I can't understand what I did wrong.
However, when I have a look at the execution stack, I always get a message (written in green, so it doesn't look like an error) like : 
> Job xx : 
> Reason : Missing output files : 

And finally, is it normal that we get Job 17, then Job 12, then Job 2, then Job 14, etc. ?

Could someone help me please ? Thanks in advance !

**Comment by NJeanray** [comment]
Hello,

Can anyone help regarding this issue ?

Best regards,
Nathalie

**Comment by lucventurini** [comment]
Dear @NJeanray 

Apologies for the delay. I am now back working almost full time on Mikado.

The easiest explanation is that the file system lags in showing that the files have been created to the Snakemake process.

Would you please be able to relaunch Daijin with the extra switch:

`--latency-wait 120`

(2 minutes)? If it does not function, we will have a deeper look at the code.

**Comment by NJeanray** [comment]
Dear @lucventurini,

Thank you for your answer, do not worry regarding the delay.

I tried your suggestion, and I added the extra switch:

`daijin assemble --cores 40 --latency-wait 120 daijin.yaml`

But unfortunately, I still get an error:

```
[Fri Oct  9 09:49:52 2020]
Job 10: Using portcullis to filter invalid junctions: hisat-Patient_24-0
Reason: Missing output files: Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.bed, Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.tab; Input files updated by another job: Daijin/4-portcullis/portcullis_hisat-Patient_24-0/2-junc/hisat-Patient_24-0.junctions.tab

  portcullis filter -o Daijin/4-portcullis/portcullis_hisat-Patient_24-0/3-filt/hisat-Patient_24-0 --canonical=C,S --max_length=10000  --threads=4 Daijin/4-portcullis/portcullis_hisat-Patient_24-0/1-prep Daijin/4-portcullis/portcullis_hisat-Patient_24-0/2-junc/hisat-Patient_24-0.junctions.tab > Daijin/4-portcullis/logs/hisat-Patient_24-0/filter.log 2>&1 && ln -sf ../portcullis_hisat-Patient_24-0/3-filt/hisat-Patient_24-0.pass.junctions.bed Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.bed || ln -sf ../portcullis_hisat-Patient_24-0/2-junc/hisat-Patient_24-0.junctions.bed Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.bed &&  ln -sf ../portcullis_hisat-Patient_24-0/3-filt/hisat-Patient_24-0.pass.junctions.tab Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.tab || ln -sf ../portcullis_hisat-Patient_24-0/2-junc/hisat-Patient_24-0.junctions.tab Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.tab && touch -h Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.bed && touch -h Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.tab
Waiting at most 120 seconds for missing files.
MissingOutputException in line 997 of /softs/anaconda3/lib/python3.7/site-packages/Mikado/daijin/assemble.smk:
Job completed successfully, but some output files are missing. Missing files after 120 seconds:
Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.tab
This might be due to filesystem latency. If that is the case, consider to increase the wait time with --latency-wait.
  File "/softs/anaconda3/lib/python3.7/site-packages/snakemake/executors/__init__.py", line 544, in handle_job_success
  File "/softs/anaconda3/lib/python3.7/site-packages/snakemake/executors/__init__.py", line 231, in handle_job_success
Removing output files of failed job portcullis_filter since they might be corrupted:
Daijin/4-portcullis/output/portcullis_hisat-Patient_24-0.pass.junctions.bed
[Fri Oct  9 09:51:57 2020]
Finished job 13.
11 of 20 steps (55%) done
[Fri Oct  9 09:56:12 2020]
Finished job 12.
12 of 20 steps (60%) done
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /local_scratch/home/NJEANRAY/patient_24/.snakemake/log/2020-10-09T072525.442761.snakemake.log
```

Do you think I should increase the latency value ?

Thanks in advance,

Best regards

**Comment by lucventurini** [comment]
Dear @NJeanray 

Many thanks for trying my suggestion; now we know the issue is not with the filesystem. Let's try to debug this!

Could you please:

- check whether the missing output file actually exists, by following the link and checking its size (if it is 0, then things have gone wrong).
- Attach here the log file for portcullis (`Daijin/4-portcullis/logs/hisat-Patient_24-0/filter.log`)

This would allow us to understand whether there is an issue in Portcullis.

CC'ing @ljyanesm as he has worked on similar code recently.

Many thanks

**Comment by NJeanray** [comment]
Dear @lucventurini ,

Thanks again for your help and your time.

So here is the dir content:

> (base) NJEANRAY@558-openchem:/local_scratch/home/NJEANRAY/patient_24/Daijin/4-portcullis/output$ ls -lrt
> total 4
> lrwxrwxrwx 1 NJEANRAY Utilisa. du domaine 77 Oct  9 09:49 portcullis_hisat-Patient_24-0.pass.junctions.tab -> ../portcullis_hisat-Patient_24-0/3-filt/hisat-Patient_24-0.pass.junctions.tab

And when I follow the target of the symbolic link, there is no file at all:

> (base) NJEANRAY@558-openchem:/local_scratch/home/NJEANRAY/patient_24/Daijin/4-portcullis/portcullis_hisat-Patient_24-0/3-filt$ ls -lrt
> total 0

Please, find below the content of file filter.log as requested:

> (base) NJEANRAY@558-openchem:/local_scratch/home/NJEANRAY/patient_24/Daijin/4-portcullis/logs/hisat-Patient_24-0$ cat filter.log
> Portcullis V1.1.2
> 
> Running portcullis in junction filter mode
> ------------------------------------------
> 
> Loading junctions from Daijin/4-portcullis/portcullis_hisat-Patient_24-0/2-junc/hisat-Patient_24-0.junctions.tab ... done.
> Found 116900 junctions.
> 
> Self training mode activated.
> 
> Loading input junctions ...
> Portcullis junction filter completed.
> Total runtime: 4.7s
> 
> ../lib/include/portcullis/python_helper.hpp(167): Throw in function void portcullis::PyHelper::execute(std::__cxx11::string, int, char**)
> Dynamic exception type: boost::exception_detail::clone_impl<portcullis::PortcullisPythonException>
> std::exception::what: std::exception
> [portcullis::PortcullisPythonError*] = Unexpected python error
> Traceback (most recent call last):
>   File "//softs/anaconda3/share/portcullis/scripts/portcullis/rule_filter.py", line 392, in <module>
>     if __name__=='__main__': main()
>   File "//softs/anaconda3/share/portcullis/scripts/portcullis/rule_filter.py", line 385, in main
>     create_training_sets(args)
>   File "//softs/anaconda3/share/portcullis/scripts/portcullis/rule_filter.py", line 127, in create_training_sets
>     original = DataFrame.from_csv(args.input, sep='\t', header=0)
> AttributeError: type object 'DataFrame' has no attribute 'from_csv'

Many thanks !

**Comment by lucventurini** [comment]
Dear @NJeanray 

OK, this makes sense. The problem is that the installed version of `portcullis` is old and does not have a critical bugfix. 

I just checked the current code (v1.2.2) and that should function correctly. 

Would you please be able to install the latest version (e.g. through conda, see here: https://anaconda.org/bioconda/portcullis) and retry?

Kind regards

**Comment by NJeanray** [comment]
Dear @lucventurini,

I just followed your instructions and I installed a more recent version of Portcullis through Conda : Portcullis 1.1.2

Unfortunately it seems that I get the same error. Please find filter.log:

> 
> (/local_scratch/home/NJEANRAY/mikado) NJEANRAY@555-openchem:/local_scratch/home/NJEANRAY/patient_24/Daijin/4-portcullis/logs/hisat-Patient_24-0$ cat filter.log
> Portcullis V1.1.2
> 
> Running portcullis in junction filter mode
> ------------------------------------------
> 
> Loading junctions from Daijin/4-portcullis/portcullis_hisat-Patient_24-0/2-junc/hisat-Patient_24-0.junctions.tab ... done.
> Found 116900 junctions.
> 
> Self training mode activated.
> 
> Loading input junctions ...
> Portcullis junction filter completed.
> Total runtime: 4.7s
> 
> ../lib/include/portcullis/python_helper.hpp(167): Throw in function void portcullis::PyHelper::execute(std::__cxx11::string, int, char**)
> Dynamic exception type: boost::exception_detail::clone_impl<portcullis::PortcullisPythonException>
> std::exception::what: std::exception
> [portcullis::PortcullisPythonError*] = Unexpected python error
> Traceback (most recent call last):
>   File "//local_scratch/home/NJEANRAY/mikado/share/portcullis/scripts/portcullis/rule_filter.py", line 392, in <module>
>     if __name__=='__main__': main()
>   File "//local_scratch/home/NJEANRAY/mikado/share/portcullis/scripts/portcullis/rule_filter.py", line 385, in main
>     create_training_sets(args)
>   File "//local_scratch/home/NJEANRAY/mikado/share/portcullis/scripts/portcullis/rule_filter.py", line 127, in create_training_sets
>     original = DataFrame.from_csv(args.input, sep='\t', header=0)
> AttributeError: type object 'DataFrame' has no attribute 'from_csv'

Do you have an idea ?

Thanks again,

Best regards

**Comment by NJeanray** [comment]
Dear @lucventurini 

I've double checked with a collegue from empty environments and we both encountered the same issues, installing conda environment, Mikado:

`pip3 install mikado==2.0rc2 `

Removing old version of Portcullis and install it again: 

`conda install -c bioconda portcullis`

, and despite this, we're still in Portcullis version 1.1.2.

We thought that maybe it could come from the fact that we are working on a DGX station, so we've tested on a single server, but the issue is the same. We don't achieve to install version 1.2.2.

Here is  our settings, if it can help to understand what's happening:

> 
> (/local_scratch/home/NJEANRAY/mikado) NJEANRAY@555-openchem:/local_scratch/home/NJEANRAY/mikado/include$ mikado --version
> /local_scratch/home/NJEANRAY/anaconda3/lib/python3.8/site-packages/Bio/SubsMat/__init__.py:126: BiopythonDeprecationWarning: Bio.SubsMat has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.substitution_matrices as a replacement, and contact the Biopython developers if you still need the Bio.SubsMat module.
>   warnings.warn(
> Mikado v2.0rc2

> (/local_scratch/home/NJEANRAY/mikado) NJEANRAY@555-openchem:/local_scratch/home/NJEANRAY/mikado/include$ python --version
> Python 3.8.3

> (/local_scratch/home/NJEANRAY/mikado) NJEANRAY@555-openchem:/local_scratch/home/NJEANRAY/mikado/include$ portcullis --version
> portcullis 1.1.2

> (/local_scratch/home/NJEANRAY/mikado) NJEANRAY@555-openchem:/local_scratch/home/NJEANRAY/mikado/include$ conda --version
> conda 4.8.5

Best regards

**Comment by lucventurini** [comment]
Dear @NJeanray 

OK,  I think that the problem is on my end, namely, we should update the version of Mikado in `conda`. So thank you for marking this.

Possible solutions, in the meantime:

1. Create an updated `mikado` environment:
```
git clone https://github.com/EI-CoreBioinformatics/mikado
cd mikado
conda env create -n mikado2 -f environment.yml
conda activate mikado2
python setup.py bdist_wheel
pip install dist/*whl
conda install -c bioconda -y portcullis
```
2. As solution 1, but instead of installing portcullis in the environment, run `daijin` with the `--use-conda` switch.
3. Install portcullis through Docker or Singularity.
4. Compile and install `portcullis` from source.

Apologies for the incompatibility. This flew under our radar - we have been trying to push for a stable 2.0 for a while now. Once that is done, the `conda` programs will be compatible.

**Comment by NJeanray** [comment]
Dear @lucventurini,

Thank you very much !
I installed Mikado as you describe it in solution 1. Then I was able to advance in the pipeline execution.
However, I'm encountering now the following error:

> [Wed Oct 14 07:40:34 2020]
> Job 11: Using samtools to collected stats for: Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam.bai
> Reason: Missing output files: Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam.stats; Input files updated by another job: Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam, Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam.bai
> 
>  samtools stats Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam > Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam.stats && plot-bamstats -p Daijin/2-alignments/output/plots/hisat-Patient_24-0/hisat-Patient_24-0 Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam.stats
> Activating conda environment: /local_scratch/home/NJEANRAY/patient_24/.snakemake/conda/b300b8d4
> gnuplot: error while loading shared libraries: libGL.so.1: cannot open shared object file: No such file or directory
> The command exited with non-zero status 32512:
>         gnuplot Daijin/2-alignments/output/plots/hisat-Patient_24-0/hisat-Patient_24-0-quals.gp
> 
>  at /local_scratch/home/NJEANRAY/patient_24/.snakemake/conda/b300b8d4/bin/plot-bamstats line 51.
>         main::error("The command exited with non-zero status 32512:\x{a}\x{9}gnuplot Daiji"...) called at /local_scratch/home/NJEANRAY/patient_24/.snakemake/conda/b300b8d4/bin/plot-bamstats line 317
>         main::plot("Daijin/2-alignments/output/plots/hisat-Patient_24-0/hisat-Pat"...) called at /local_scratch/home/NJEANRAY/patient_24/.snakemake/conda/b300b8d4/bin/plot-bamstats line 653
>         main::plot_qualities(HASH(0x5620617c03e8)) called at /local_scratch/home/NJEANRAY/patient_24/.snakemake/conda/b300b8d4/bin/plot-bamstats line 33
> [Wed Oct 14 07:43:46 2020]
> Error in rule bam_stats:
>     jobid: 11
>     output: Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam.stats, Daijin/2-alignments/output/plots/hisat-Patient_24-0/hisat-Patient_24-0-quals.png
>     conda-env: /local_scratch/home/NJEANRAY/patient_24/.snakemake/conda/b300b8d4
>     shell:
>          samtools stats Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam > Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam.stats && plot-bamstats -p Daijin/2-alignments/output/plots/hisat-Patient_24-0/hisat-Patient_24-0 Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam.stats
>         (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)
> 
> Removing output files of failed job bam_stats since they might be corrupted:
> Daijin/2-alignments/output/hisat-Patient_24-0.sorted.bam.stats
> Shutting down, this might take some time.
> Exiting because a job execution failed. Look above for error message
> Complete log: /local_scratch/home/NJEANRAY/patient_24/.snakemake/log/2020-10-14T062111.640682.snakemake.log

Please, find the mentionned log file: [2020-10-14T062111.640682.snakemake.log](https://github.com/EI-CoreBioinformatics/mikado/files/5376701/2020-10-14T062111.640682.snakemake.log)

Also, if you prefer me to open an new issue regarding this point, do not hesitate to tell me.

Thanks again,

Best regards

**Comment by lucventurini** [comment]
Dear @NJeanray 

Glad to hear that the first solution worked!

As for the new issue, the error is most probably caused by gnuplot, the bioconda version often gives me trouble.

Could you try:

```
conda install -c conda-forge -y -n mikado2 -- conda-forge::gnuplot
```

and retry the pipeline?

Kind regards

**Comment by NJeanray** [comment]
Dear @lucventurini,

I did what you adviced, but I still get the same error.

Best regards
